server/service/src/sensor/berlinger.rs

```diff
@@ -0,0 +1,380 @@
+            sensor_id: sensor_row.id.clone(),
+            location_id: sensor_row.location_id.clone(),
+            temperature: temperature_log.temperature,
+            datetime: temperature_log.timestamp,
+            temperature_breach_id: None,
+        };
+        TemperatureLogRowRepository::new(connection).upsert_one(&new_temperature_log)?;
+        println!("Added sensor log {:?} ", new_temperature_log);
+        Ok(())
+    }
+}
+
+fn sensor_add_breach_if_new(
+    connection: &StorageConnection,
+    sensor_row: &SensorRow,
+    temperature_breach: &temperature_sensor::TemperatureBreach,
+    breach_config: &temperature_sensor::TemperatureBreachConfig,
+) -> Result<(), RepositoryError> {
+    let breach_row_type = get_breach_row_type(&temperature_breach.breach_type);
+    let result = get_matching_sensor_breach(
+        connection,
+        &sensor_row.id,
+        temperature_breach.start_timestamp,
+        temperature_breach.end_timestamp,
+        &breach_row_type,
+    )?;
+
+    if let Some(mut record) = result.clone().pop() {
+        if record.temperature_breach_row.end_datetime != Some(temperature_breach.end_timestamp) { // Update breach end time if it has changed
+            record.temperature_breach_row.end_datetime = Some(temperature_breach.end_timestamp);
+            TemperatureBreachRowRepository::new(connection).upsert_one(&record.temperature_breach_row)?;
+        }
+        Ok(())
+    } else {
+        let new_temperature_breach = TemperatureBreachRow {
+            id: uuid(),
+            store_id: sensor_row.store_id.clone(),
+            sensor_id: sensor_row.id.clone(),
+            location_id: sensor_row.location_id.clone(),
+            start_datetime: temperature_breach.start_timestamp,
+            end_datetime: Some(temperature_breach.end_timestamp),
+            acknowledged: false,
+            duration: temperature_breach.duration.num_seconds() as i32,
+            r#type: breach_row_type,
+            threshold_duration: breach_config.duration.num_seconds() as i32,
+            threshold_minimum: breach_config.minimum_temperature,
+            threshold_maximum: breach_config.maximum_temperature,
+        };
+        TemperatureBreachRowRepository::new(connection).upsert_one(&new_temperature_breach)?;
+        println!("Added sensor breach {:?} ", new_temperature_breach);
+        Ok(())
+    }
+}
+
+fn sensor_add_breach_config_if_new(
+    connection: &StorageConnection,
+    sensor_row: &SensorRow,
+    temperature_breach_config: &temperature_sensor::TemperatureBreachConfig,
+) -> Result<(), RepositoryError> {
+    let mut config_description = format!(
+        "for {} minutes",
+        temperature_breach_config.duration.num_minutes()
+    );
+    let breach_row_type = get_breach_row_type(&temperature_breach_config.breach_type);
+
+    match temperature_breach_config.breach_type {
+        BreachType::ColdConsecutive => {
+            config_description = format!(
```

Comment:
Actually do we need to store this in description, can we deduce matching config by it's configuration ?

Reply:
Yes, we could match by its config but that's four fields to filter on. The config description is set as a unique field in the DB, and it isn't set in the Berlinger data => generating it like this creates a human-readable unique description

Reply:
If it's human readable, then we need to consider translations, I think it becomes more complex at this point ? I would assume anywhere we would show breach config in the UI we will show enough information to read breach config (without description). Also I think when we have manually configured breaches, use would be able to set their own description

> but that's four fields to filter on.

What wondering why it's a `but`, would it be complex or a lot of code ?.


I think we should match on the fields, but doesn't stop this PR, we can do in another issue